source/common/ssl/context_impl.cc

```diff
@@ -578,163 +607,176 @@ ServerContextImpl::ServerContextImpl(Stats::Scope& scope, const ServerContextCon
   if (config.tlsCertificates().empty()) {
     throw EnvoyException("Server TlsCertificates must have a certificate specified");
   }
-  if (config.certificateValidationContext() != nullptr &&
-      !config.certificateValidationContext()->caCert().empty()) {
-    bssl::UniquePtr<BIO> bio(
-        BIO_new_mem_buf(const_cast<char*>(config.certificateValidationContext()->caCert().data()),
-                        config.certificateValidationContext()->caCert().size()));
-    RELEASE_ASSERT(bio != nullptr, "");
-    // Based on BoringSSL's SSL_add_file_cert_subjects_to_stack().
-    bssl::UniquePtr<STACK_OF(X509_NAME)> list(sk_X509_NAME_new(
-        [](const X509_NAME** a, const X509_NAME** b) -> int { return X509_NAME_cmp(*a, *b); }));
-    RELEASE_ASSERT(list != nullptr, "");
-    for (;;) {
-      bssl::UniquePtr<X509> cert(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr));
-      if (cert == nullptr) {
-        break;
+  // First, configure the base context for ClientHello interception.
+  // TODO(htuch): replace with SSL_IDENTITY when we have this as a means to do multi-cert in
+  // BoringSSL.
+  SSL_CTX_set_select_certificate_cb(
+      tls_contexts_[0].ssl_ctx_.get(),
+      [](const SSL_CLIENT_HELLO* client_hello) -> ssl_select_cert_result_t {
+        return static_cast<ServerContextImpl*>(
+                   SSL_CTX_get_app_data(SSL_get_SSL_CTX(client_hello->ssl)))
+            ->selectTlsContext(client_hello);
+      });
+  for (auto& ctx : tls_contexts_) {
```

Comment:
Hmm, I think my reply was lost here. For both (1), we need to create fresh resources, since the consuming code requires a `.release()` and ownership transfer. So, we need to rebuild these X509 stack objects. For (2), the server cert CN is included in the hash, so also needs to be done for each TlsContext.

However, I agree that this loop was way too massive. I've refactored the code and it should be more readable, PTAL.

Reply:
re (1) I think you could use `bssl::UpRef()`, since the underlying objects are ref-counted.

re (2) Yes, but you can include CNs from all certificates in a single hash that's going to be used by this `ServerContextImpl`, you don't need to have separate session context ID for each certificate. Sorry if that wasn't clear in my original comment.

Reply:
For (1), if you look at the implementation of  `SSL_CTX_set_client_CA_list`, it's physically mutating the `STACK_O`F by performing a pop operation. I think what we should do is keep this one as is; even if we could ref count, IMHO it's not worth it in terms of code clarity.

I'll look at (2), this seems reasonable, thanks.

Reply:
Sounds good, thanks.